Specify subtyping requirements (#391)
ferrocene/specification: src/types-and-traits.rst

   #. :dp:`fls_9ae3idezsths`
      It is a static error if :t:`variance` ``V`` is ``any``.

:dp:`fls_WD6uyTCziRac`
:t:`[Expression]s` and :t:`[statement]s` may impose :t:`subtyping` requirements

> on their subexpressions. Such requirements are applied after

Reviewer: ```suggestion
on their :t:`[subexpression]s`. Such requirements are applied after
```

> :t:`type inference`, on the inferred :t:`[type]s` of the respective

Reviewer: ```suggestion
:t:`type inference`, on the :t:`[inferred type]s` of the respective
```

:t:`[expression]s` and :t:`[pattern]s`.

:dp:`fls_xURR0owesaIE`
It is a static error if any :t:`subtyping` requirements are not met.

:dp:`fls_CvZNYIfnOUcc`
The :t:`subtyping` requirements for :t:`[statement]s` are as follows:

* :dp:`fls_xrLYQX7W9OnR`
  :t:`[Item statement]s` impose no additional :t:`subtyping` requirements.

* :dp:`fls_DWPN8DRysgMa`
  :t:`[Let statement]s` require that the :t:`type` of the :t:`expression` of the
  :t:`let initializer` (if any) is a :t:`subtype` of the :t:`type` of the
  :t:`let statement`'s :t:`pattern`.

* :dp:`fls_tOn5oKtp300J`
  :t:`[Expression statement]s` impose the :t:`subtyping` requirements for the
  contained :t:`expression`, as outlined below.

:dp:`fls_HqXTVi7gr9wR`
The :t:`subtyping` requirements for :t:`[expression]s` are as follows:

* :dp:`fls_lxfPvK7NDNlj`
  The requirements for any :t:`arithmetic expression`, :t:`bit expression`,
  :t:`comparison expression`, :t:`compound assignment expression`,

>   :t:`negation expression` or :t:`index expression` are the same requirements

Reviewer: ```suggestion
  :t:`index expression`, or :t:`negation expression` are the same requirements
```
Just being OCD. :D
